### PR TITLE
JAMES-3753 Fix FlowedMessageUtils.deflow()

### PR DESCRIPTION
**Pull request by cketti**
Jira: https://issues.apache.org/jira/browse/JAMES-3753

### CODE REVIEW

**Review by chibenwa — APPROVED** [review]
Good work!

For complex string parsing edge cases I did find fuzzing to be incredibly effective at fiabilizing things. Things like the jazzer fuzzer https://github.com/CodeIntelligenceTesting/jazzer are easy to set up and very effective at finding such bugs!

**Comment by chibenwa** [comment]
Ok so the following test:

```
public class FlowFuzz {
    public static void fuzzerTestOneInput(FuzzedDataProvider input) {
        final String input1 = input.consumeRemainingAsString();
        try {
            FlowedMessageUtils.flow(input1, false);
        } catch (Throwable t) {
            System.out.println(Base64.getEncoder().encodeToString(input1.getBytes(StandardCharsets.UTF_8)));
        }
    }
}
```

With the following input (base64 encoded)

```
$ base64 oom-b0a8859b8f75555d874226039f27ab9273be64ed 
Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+
Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+//////////////////////8oCiAAAAAKIAo+AAAA
AAABAAAAAD4gCiAAAAAAAAAAAAAAXAAAAAAAAAAAAQAAAAAACj4gCiAKIAogCiAKIAogCiAKIAog
CiAKCiAKIAogCiAKKAo+ICA+0N/VCgoKPiAACgoKCj4gCgo+Pj5dPisKPiAgPtDf1VsKCj4gAAoK
Cgo+IP//////AAAKCgoKPiAKCj4+Pl0+IAoAKwo+ICA+0N/VWwoKPiAACgoKCj4gCv//////Gxsb
GxsbGxsAAAAAAAAAAAAAAAAjCiAKPgogCiAAAAAAAAAxAAAACh0dHR0dHQAAAAAAAAAAAAAABgA=
```

Yields an out of memory exception...

Usual suspect: no progress done in a while loop, effectively creating an infinite loop. I would start looking at FlowedMessageUtils::flow line 223.

Would it be ok to you to have a look at this?

Regards.

**Comment by cketti** [comment]
I can't reproduce this. The following test runs fine for me:

```java
@Test
void oomTest() {
    String encodedInput = "Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4AAAA=";
    String input = new String(Base64.getDecoder().decode(encodedInput), Charsets.UTF_8);

    String result = FlowedMessageUtils.deflow(input, DEL_SP_NO);

    assertThat(result).isEqualTo(">>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>> \u0000\u0000\u0000");
}
```

The loop code looks okay to me, too. 

In general, the `deflow()` code is using a lot of allocations that could be avoided. Rewriting the code is on my ToDo list. But the change in this PR shouldn't make the code any worse than it was before.

**Comment by chibenwa** [comment]
The OOM is on flow, not deflow.

This can IMO be hadnled in a separate PR/ticket if needed.

```
    @Test
    void oomTest() {
        String encodedInput = "Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4AAAA=";
        String input = new String(Base64.getDecoder().decode(encodedInput), Charsets.UTF_8);

        String result = FlowedMessageUtils.flow(input, true);
    }
```

Yields

```
Exception in thread "main" java.lang.OutOfMemoryError: Java heap space
	at java.base/java.util.Arrays.copyOf(Arrays.java:3745)
	at java.base/java.lang.AbstractStringBuilder.ensureCapacityInternal(AbstractStringBuilder.java:172)
	at java.base/java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:633)
	at java.base/java.lang.StringBuilder.append(StringBuilder.java:217)
	at org.apache.mailet.base.FlowedMessageUtils.flow(FlowedMessageUtils.java:254)
	at org.apache.mailet.base.FlowedMessageUtils.flow(FlowedMessageUtils.java:198)
	at org.apache.mailet.base.FlowedMessageUtilsTest.oomTest(FlowedMessageUtilsTest.java:17)
```

**Comment by cketti** [comment]
Oh, I see. We don't use the `flow()` method in K-9 Mail, so in my mind it didn't exist :grinning: 

The problem is that the code assumes the length of the quote prefix never exceeds the maximum target width for a line (`RFC2646_WIDTH` - 78 characters). 

This bug can be fixed by changing the line

https://github.com/apache/james-project/blob/a6f6289562c5ed42fd7dbf8cbf8a6cb972b3a2b5/mailet/base/src/main/java/org/apache/mailet/base/FlowedMessageUtils.java#L247

to 

```java
j = Math.max(width - 1, extra);
```

I think that should be a separate PR. I'm happy to open one once this change has been merged.

**Review by Arsnael — APPROVED** [review]

File: mailet/base/src/test/java/org/apache/mailet/base/FlowedMessageUtilsTest.java [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+        String input = "Text that should be \r\n" +
+            "displayed on one line";
+
+        String result = FlowedMessageUtils.deflow(input, DEL_SP_NO);
+
+        assertThat(result).isEqualTo("Text that should be displayed on one line");
+    }
+
+    @Test
+    void deflowWwithOnlySomeLinesEndingInSpace() {
```
Comment (chibenwa) [inline mailet/base/src/test/java/org/apache/mailet/base/FlowedMessageUtilsTest.java:41]:
```suggestion
    void deflowWithOnlySomeLinesEndingInSpace() {
```

**Comment by chibenwa** [comment]
> I think that should be a separate PR. I'm happy to open one once this change has been merged.

:+1:  I let you open it then.

**Comment by chibenwa** [comment]
One more defect for `flow`:

```
#20255	REDUCE cov: 47 ft: 241 corp: 58/875b lim: 80 exec/s: 0 rss: 169Mb L: 20/63 MS: 1 EraseBytes-

== Java Exception: java.lang.StringIndexOutOfBoundsException: String index out of range: 77
	at java.base/java.lang.StringLatin1.charAt(StringLatin1.java:47)
	at java.base/java.lang.String.charAt(String.java:693)
	at org.apache.mailet.base.FlowedMessageUtils.isAlphaChar(FlowedMessageUtils.java:342)
	at org.apache.mailet.base.FlowedMessageUtils.flow(FlowedMessageUtils.java:242)
	at org.apache.mailet.base.FlowedMessageUtils.flow(FlowedMessageUtils.java:198)
	at Flow.fuzzerTestOneInput(Flow.java:26)
DEDUP_TOKEN: 98734b6baa46951b
== libFuzzer crashing input ==
MS: 3 ChangeByte-EraseBytes-InsertRepeatedBytes-; base unit: ee65a29d115f109f598e8723b7d80c9adb0e24bc
0x3e,0x3e,0x3e,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x0,0x3e,0x3e,0x3e,
>>>\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00>>>
artifact_prefix='./'; Test unit written to ./crash-543b5ca62bb8cc47cd7a43b021375c0343d07ae1
Base64: Pj4+AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA+Pj4=
```

**Comment by chibenwa** [comment]
And yet another one...

```
#161909	NEW    cov: 80 ft: 553 corp: 169/16Kb lim: 589 exec/s: 17989 rss: 279Mb L: 309/540 MS: 1 PersAutoDict- DE: "a\x00\x00\x00"-

== Java Exception: java.lang.IndexOutOfBoundsException: start 0, end 85, length 84
	at java.base/java.lang.AbstractStringBuilder.checkRange(AbstractStringBuilder.java:1716)
	at java.base/java.lang.AbstractStringBuilder.append(AbstractStringBuilder.java:631)
	at java.base/java.lang.StringBuilder.append(StringBuilder.java:217)
	at org.apache.mailet.base.FlowedMessageUtils.flow(FlowedMessageUtils.java:254)
	at org.apache.mailet.base.FlowedMessageUtils.flow(FlowedMessageUtils.java:198)
	at Flow.fuzzerTestOneInput(Flow.java:26)
DEDUP_TOKEN: f8ed07a58079a9ed
== libFuzzer crashing input ==
MS: 5 CopyPart-ChangeBinInt-EraseBytes-CrossOver-EraseBytes-; base unit: 2918d0c256604a4173fc474a5991c0c7d0c76d3e
artifact_prefix='./'; Test unit written to ./crash-fa906a86e550aa4c7c18706786622281dc55877f
reproducer_path='.'; Java reproducer written to ./Crash_fa906a86e550aa4c7c18706786622281dc55877f.java
```

```
$ base64 ./crash-fa906a86e550aa4c7c18706786622281dc55877f
PgAAAAAAAAAACQA+S09LS0v///8BRQAAAAAA/yAKinJvbSAgCkZyb20gIApGcm9tIH8AAABtICAK
RnJvbSBtICAKPj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+
Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4KRnJvbSB/AAAAbSAgAAAAAAAAAAAK
RnJvbSAgACAgCkZyZ20gIAogCoogCkZyAAAAAP//S0tLb20gIApvbSAgCkZyb20gIAoKIApGcm+w
ICAKRnJvbSAgRnJvbSAKRnJvbSAgACBGcktvbSAgACBLSwoAAAAAAABLS0tLS0tLS0tLS0s+IAJr
S0tLS0tLS0tLIAogCnAKQwBGODg4cm9tICAKRnJvbSAgRnI4ODg4ODg4ODg4ODhLODg4ODg4ODg4
ODg4ODhLODg4OG9tICAKRnJvbSAgACBLSxcXUzg4OAogCgoKIAoKIApGcm9tICAAIAo4ODg4ODg4
ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg4ODg+
PgAAAAA=
```

**Comment by chibenwa** [comment]
And yet another one...

```
#58742	REDUCE cov: 80 ft: 495 corp: 138/8018b lim: 198 exec/s: 58742 rss: 278Mb L: 85/198 MS: 4 ChangeByte-EraseBytes-ChangeBit-CopyPart-

== Java Exception: java.lang.StringIndexOutOfBoundsException: String index out of range: -1
	at java.base/java.lang.String.substring(String.java:1841)
	at org.apache.mailet.base.FlowedMessageUtils.flow(FlowedMessageUtils.java:262)
	at org.apache.mailet.base.FlowedMessageUtils.flow(FlowedMessageUtils.java:198)
	at Flow.fuzzerTestOneInput(Flow.java:26)
DEDUP_TOKEN: aa788d652efbfb4b
== libFuzzer crashing input ==
MS: 2 EraseBytes-CopyPart-; base unit: 1a3229fda9b1c877a54447de84a885dc7e7838d2
0xa9,0x1,0x0,0x0,0xa,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0xa9,0x1,0x0,0x0,0xa,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0x3e,0xa,
\xa9\x01\x00\x00\x0a>>>>>>>>>>>>>>>>>>>>>>>>>>>>\xa9\x01\x00\x00\x0a>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>>\x0a
artifact_prefix='./'; Test unit written to ./crash-f6a4239a7a187c27130d1704cb9bcd2da1ebde2e
Base64: qQEAAAo+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+qQEAAAo+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Pj4+Cg==
```

**Comment by chibenwa** [comment]
And yet another one...

```
#109	REDUCE cov: 31 ft: 71 corp: 9/24b lim: 4 exec/s: 0 rss: 129Mb L: 2/4 MS: 2 ChangeBit-EraseBytes-



==201969== ERROR: libFuzzer: out-of-memory (used: 2573Mb; limit: 2048Mb)
   To change the out-of-memory limit use -rss_limit_mb=<N>

MS: 3 ChangeBinInt-CrossOver-CMP- DE: "\x00>"-; base unit: 11d7d7ce0b68973d37325a5b3f1fdcf7d2e88954
0x0,0x3e,
\x00>
artifact_prefix='./'; Test unit written to ./oom-9d394f37e0e251ccbe3c477f6bb122005c38f167
Base64: AD4=
```

**Pull request merged**